### PR TITLE
On followers, replay events in batches

### LINKED ISSUE #7597: On followers, be aware of batches when creating a snapshot
**Description**

When the stream processor processes a command then it writes one or more events and follow-up commands. All state changes that are reflected by the events are written within one transaction in RocksDB. If a snapshot is taken then it contains all state changes.

On the follower, it's a bit different. The follower reads all events one by one and applies the state changes. The state changes of one event are written whin one transaction. But multiple events can have the same source record (i.e. are produced by the same command).  

If a snapshot is taken on the follower then it may contain only some of the state changes but not from all events that are produced by a command.  

As a result, the next replay may skip events or apply them again. It is related to the `lastProcessedPosition` that is stored in the state. On replay, the position is used as the point to start reading from, and to determine if the event is already applied.

One solution could be to write all state changes that have the same source record within one transaction. After the last event is replayed (i.e. read an event with a different source record, or reach the end of the log), the transaction is committed.

### PR DESCRIPTION
**Pull request by saig0**
## Description

* apply all events of the same command within one database transaction
* ensure that the last processed position is set only after all events are applied

Note that the PR adds no new test cases because the transactional behavior is hard to test. However, there are tests for the batch reader to ensure the behavior.

## Related issues

closes #7597 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java [hunk 1]
```diff
@@ -164,28 +171,41 @@ void replayNextEvent() {
       }
 
     } catch (final RuntimeException e) {
-      final var replayException =
-          new ProcessingException("Unable to replay record", currentEvent, metadata, e);
-      recoveryFuture.completeExceptionally(replayException);
+      recoveryFuture.completeExceptionally(new RuntimeException("Failed to replay records", e));
```
Comment (deepthidevaki) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java:174]: Can you add more context to the error message? Such as lastSourceEventPostion, lastReplayedPosition etc.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java [hunk 2]
```diff
@@ -164,28 +171,41 @@ void replayNextEvent() {
...
     }
 
-    readMetadata(currentEvent);
-    final var currentTypedEvent = readRecordValue(currentEvent);
+    zeebeDbTransaction = transactionContext.getCurrentTransaction();
+    zeebeDbTransaction.run(
+        () -> {
+          batch.forEachRemaining(this::replayEvent);
+
+          lastProcessedPositionState.markAsProcessed(lastSourceEventPosition);
```
Comment (deepthidevaki) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java:191]: :thought_balloon: This atomically updates the lastProcessedPosition in the state. But for snapshotting, we use the inmemory `lastSourceEventPosition` (`StreamProcessor:getLastProcessedPositionAsync`) which is not updated atomically after a batch is processed. Instead it is updated after each event is replayed. Would this be a problem? It is used to decide which position can be compacted.

**Review by deepthidevaki — COMMENTED** [review]
Looks good. I have some questions. Please see my comments.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java (hunk 2 above)
Comment (saig0) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/ReplayStateMachine.java:191]: Good point. It's better to set the position after the transaction is committed :+1:

**Comment by saig0** [comment]
@deepthidevaki I applied your hints. Please have another look :cookie:

**Review by deepthidevaki — APPROVED** [review]
:100: I think now we can consistency build state on followers :slightly_smiling_face:

**Pull request merged**